### PR TITLE
Create TreeMenu component

### PR DESCRIPTION
**Pull request by dzonidoo**
SDESK-7040

### CODE REVIEW

File: app-typescript/components/TreeSelect/TreeSelectItem.tsx [hunk 1]
```diff
@@ -25,7 +26,10 @@ export class TreeSelectItem<T> extends React.Component<IProps<T>> {
     render() {
         return (
             <li
-                className='suggestion-item suggestion-item--multi-select'
+                className={
+                    'suggestion-item suggestion-item--multi-select'
+                    + (this.props.labelWrap ? ' suggestion-item--wrap' : '')
```
Comment (tomaskikutis) [inline app-typescript/components/TreeSelect/TreeSelectItem.tsx]: how about making it wrap unconditionally?

File: app-typescript/components/TreeSelect/TreeSelectItem.tsx [hunk 2]
```diff
@@ -55,7 +59,9 @@ export class TreeSelectItem<T> extends React.Component<IProps<T>> {
                     <span
                         className={
                             'suggestion-item--bgcolor'
-                            + (this.props.selectedItem ? ' suggestion-item--disabled' : '')}
+                            + (this.props.selectedItem ? ' suggestion-item--disabled' : '')
+                            + (this.props.labelWrap ? ' suggestion-item--bgcolor--wrap' : '')
```
Comment (tomaskikutis) [inline app-typescript/components/TreeSelect/TreeSelectItem.tsx]: is `bg-color` related to wrapping? how?

File: app-typescript/components/TreeMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+
+function nodeHasChildren<T>(item: IParent<T> | IChildren<T>): item is IParent<T> {
+    return item['children'] != null;
+}
+
+function nodeCanBeSelected<T>(item: IParent<T> | IChildren<T>): item is IChildren<T> {
+    return item['onSelect'] != null;
+}
+
+export class TreeMenu<T> extends React.Component<IProps<T>, IState<T>> {
```
Comment (tomaskikutis) [inline app-typescript/components/TreeMenu.tsx:56]: I see there's lot of code duplication between this and tree select. I expect we should now be able to reimplement tree select something like this:

```

export class TreeSelect extends React.Component {
    render() {
        return (
            <TreeMenu
                getOptions={() => options2}
                getId={(item) => item.name}
                getLabel={(item) => item.name}
                optionTemplate={(item) => {
                    return (
                        <div
                            style={{
                                display: 'flex',
                                alignItems: 'center',
                                gap: '4px'
                            }}
                        >
                            {item.icon}
                            {item.name}
                        </div>
                    )
                }}
            >
                {(toggle) => (
                    <div>
                        <div>label</div>

                        <div>
                            {
                                this.state.values.map((val, i) => (
                                    <span key={i}>{val}</span>
                                ))
                            }
                        </div>

                        <button onClick={toggle}>+</button>
                    </div>
                )}
            </TreeMenu>
        );
    }
}
```

File: app-typescript/components/TreeSelect/TreeSelectItem.tsx (hunk 1 above)
Comment (dzonidoo) [reply app-typescript/components/TreeSelect/TreeSelectItem.tsx]: but why?

File: app-typescript/components/TreeMenu.tsx (hunk 3 above)
Comment (dzonidoo) [reply app-typescript/components/TreeMenu.tsx:56]: but what about async mode on treeSelect?
treeMenu not support it

**Review by tomaskikutis — APPROVED** [review]

**Pull request merged**